File: src/tools/file-read.ts
// ============================================================================
// file_read 工具——读取文件及指定行范围
// ============================================================================

import * as fs from "node:fs/promises";
import { createHash } from "node:crypto";
import type { ITool, ToolResult, ToolExecutionContext, JSONSchema7 } from "./types.js";
import type { PermissionLevel } from "../core/types.js";
import { z } from "zod";
import {
  resolveExistingWorkspacePath,
  WorkspacePathError,
} from "./path-policy.js";

// ---------------------------------------------------------------------------
// 输入输出类型
// ---------------------------------------------------------------------------

export interface FileReadInput {
  path: string;
  startLine?: number;
  endLine?: number;
  maxBytes?: number;
}

export interface FileReadOutput {
  path: string;
  content: string;
  startLine: number;
  endLine: number;
  totalLines: number;
  truncated: boolean;
  sha256: string;
}

// ---------------------------------------------------------------------------
// 工具实现
// ---------------------------------------------------------------------------

const DEFAULT_MAX_BYTES = 2 * 1024 * 1024; // 2 MB

const fileReadInputValidator = z.object({
  path: z.string().min(1).max(4096),
  startLine: z.number().int().min(1).optional(),
  endLine: z.number().int().min(1).optional(),
  maxBytes: z.number().int().min(1).max(DEFAULT_MAX_BYTES).optional(),
}).strict().refine(
  (input) => input.endLine === undefined || input.startLine === undefined || input.endLine >= input.startLine,
  { message: "endLine 必须大于或等于 startLine", path: ["endLine"] }
);

export class FileReadTool implements ITool<FileReadInput, FileReadOutput> {
  name = "file_read";
  description =
    "读取工作区内文件，可指定行范围。超过 2 MiB 的文件会被截断，并返回 SHA-256。";
  permissionLevel: PermissionLevel = "read-only";
  defaultTimeoutMs = 30_000;
  inputValidator = fileReadInputValidator;

  inputSchema: JSONSchema7 = {
    type: "object",
    properties: {
      path: {
        type: "string",
        minLength: 1,
        maxLength: 4096,
        description: "要读取的工作区内文件路径",
      },
      startLine: {
        type: "number",
        minimum: 1,
        description: "起始行，从 1 开始，默认第 1 行",
      },
      endLine: {
        type: "number",
        minimum: 1,
        description: "结束行，包含该行，默认到文件末尾",
      },
      maxBytes: {
        type: "number",
        minimum: 1,
        maximum: DEFAULT_MAX_BYTES,
        description: "最大读取字节数，默认 2 MiB",
      },
    },
    required: ["path"],
    additionalProperties: false,
  };

  idempotencyKey(input: FileReadInput, _context: ToolExecutionContext): string {
    return `file_read:${input.path}:${input.startLine ?? 0}:${input.endLine ?? 0}`;
  }

  async execute(
    input: FileReadInput,
    context: ToolExecutionContext
  ): Promise<ToolResult<FileReadOutput>> {
    const startedAt = new Date().toISOString();
    const maxBytes = input.maxBytes ?? DEFAULT_MAX_BYTES;

    try {
      // 统一解析真实路径，避免绝对路径、.. 和符号链接逃逸工作区。
      const resolvedPath = await resolveExistingWorkspacePath(input.path, context.cwd);

      // 检查目标是否为可访问的普通文件。
      let stat: Awaited<ReturnType<typeof fs.stat>>;
      try {
        stat = await fs.stat(resolvedPath);
      } catch {
        return {
          success: false,
          output: null,
          error: {
            code: "FILE_NOT_FOUND",
            message: `File not found: ${input.path}`,
            retryable: false,
            details: { path: input.path, resolvedPath },
          },
          durationMs: 0,
          startedAt,
          endedAt: new Date().toISOString(),
          metadata: {},
        };
      }

      if (!stat.isFile()) {
        return {
          success: false,
          output: null,
          error: {
            code: "NOT_A_FILE",
            message: `Path is not a file: ${input.path}`,
            retryable: false,
            details: { path: input.path, resolvedPath },
          },
          durationMs: 0,
          startedAt,
          endedAt: new Date().toISOString(),
          metadata: {},
        };
      }

      // 按大小限制读取文件。
      let content: string;
      if (stat.size > maxBytes) {
        // 大文件只读取前 maxBytes 个字节。
        const buf = Buffer.alloc(maxBytes);
        const fd = await fs.open(resolvedPath, "r");
        try {
          await fd.read(buf, 0, maxBytes, 0);
        } finally {
          await fd.close();
        }
        content = buf.toString("utf-8");
      } else {
        content = await fs.readFile(resolvedPath, "utf-8");
      }

      const truncated = stat.size > maxBytes;
      const lines = content.split("\n");

      // 截取用户指定的行范围。
      const startLine = input.startLine ?? 1;
      const endLine = input.endLine ?? lines.length;
      const selectedLines = lines.slice(Math.max(0, startLine - 1), endLine);
      const selectedContent = selectedLines.join("\n");

      const sha256 = createHash("sha256").update(content).digest("hex");

      const endedAt = new Date().toISOString();
      return {
        success: true,
        output: {
          path: input.path,
          content: selectedContent,
          startLine,
          endLine: Math.min(endLine, lines.length),
          totalLines: lines.length,
          truncated,
          sha256,
        },
        error: null,
        durationMs: new Date(endedAt).getTime() - new Date(startedAt).getTime(),
        startedAt,
        endedAt,
        metadata: {
          fileSize: stat.size,
          resolvedPath,
        },
      };
    } catch (err: unknown) {
      const endedAt = new Date().toISOString();
      const message = err instanceof Error ? err.message : "Unknown error reading file";
      return {
        success: false,
        output: null,
        error: {
          code: err instanceof WorkspacePathError
            ? err.code
            : (err as NodeJS.ErrnoException)?.code === "ENOENT"
              ? "FILE_NOT_FOUND"
              : "FILE_READ_ERROR",
          message,
          retryable: false,
          details: { path: input.path },
        },
        durationMs: new Date(endedAt).getTime() - new Date(startedAt).getTime(),
        startedAt,
        endedAt,
        metadata: {},
      };
    }
  }
}

File: crates/xycli-core/src/tools/file_read.rs
//! `file_read`：安全读取工作区内普通文件。

use std::{path::Path, time::Duration};

use async_trait::async_trait;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use tokio::{fs, io::AsyncReadExt};

use crate::permission::PermissionLevel;

use super::path_policy::resolve_existing;
use super::{
    Tool, ToolContext, ToolDefinition, ToolResult, object, reject_unknown_fields, required_string,
};

const DEFAULT_MAX_BYTES: usize = 2 * 1024 * 1024;

pub struct FileReadTool;

#[async_trait]
impl Tool for FileReadTool {
    fn definition(&self) -> ToolDefinition {
        ToolDefinition {
            name: "file_read",
            description: "读取工作区内文件，可指定行范围。超过 2 MiB 的文件会被截断，并返回 SHA-256。",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "path": {"type":"string","minLength":1,"maxLength":4096,"description":"工作区内文件路径"},
                    "startLine": {"type":"integer","minimum":1},
                    "endLine": {"type":"integer","minimum":1},
                    "maxBytes": {"type":"integer","minimum":1,"maximum":DEFAULT_MAX_BYTES}
                },
                "required": ["path"],
                "additionalProperties": false
            }),
            permission_level: PermissionLevel::ReadOnly,
            default_timeout: Duration::from_secs(30),
        }
    }

    fn validate(&self, input: &Value) -> Result<(), Vec<String>> {
        let map = object(input)?;
        let mut issues = Vec::new();
        reject_unknown_fields(
            map,
            &["path", "startLine", "endLine", "maxBytes"],
            &mut issues,
        );
        required_string(map, "path", 4096, &mut issues);
        let start = map
            .get("startLine")
            .map(Value::as_u64)
            .transpose_option("startLine", &mut issues);
        let end = map
            .get("endLine")
            .map(Value::as_u64)
            .transpose_option("endLine", &mut issues);
        let max = map
            .get("maxBytes")
            .map(Value::as_u64)
            .transpose_option("maxBytes", &mut issues);
        if let Some(Some(value)) = start
            && value < 1
        {
            issues.push("startLine 必须大于等于 1。".into());
        }
        if let Some(Some(value)) = end
            && value < 1
        {
            issues.push("endLine 必须大于等于 1。".into());
        }
        if let (Some(Some(start)), Some(Some(end))) = (start, end)
            && end < start
        {
            issues.push("endLine 必须大于或等于 startLine。".into());
        }
        if let Some(Some(value)) = max
            && !(1..=DEFAULT_MAX_BYTES as u64).contains(&value)
        {
            issues.push(format!("maxBytes 必须在 1 到 {DEFAULT_MAX_BYTES} 之间。"));
        }
        if issues.is_empty() {
            Ok(())
        } else {
            Err(issues)
        }
    }

    async fn execute(&self, input: Value, context: ToolContext) -> ToolResult {
        let path_text = input["path"].as_str().unwrap();
        let resolved = match resolve_existing(Path::new(path_text), &context.cwd).await {
            Ok(path) => path,
            Err(error) => {
                let code = if error.message.contains("超出工作区") {
                    "PATH_OUTSIDE_WORKSPACE"
                } else {
                    "FILE_NOT_FOUND"
                };
                return ToolResult::failure(
                    code,
                    error.message,
                    context.started_at,
                    json!({"path":path_text}),
                );
            }
        };
        let metadata = match fs::metadata(&resolved).await {
            Ok(metadata) if metadata.is_file() => metadata,
            Ok(_) => {
                return ToolResult::failure(
                    "NOT_A_FILE",
                    format!("路径不是文件：{path_text}"),
                    context.started_at,
                    json!({"path":path_text}),
                );
            }
            Err(error) => {
                return ToolResult::failure(
                    "FILE_NOT_FOUND",
                    error.to_string(),
                    context.started_at,
                    json!({"path":path_text}),
                );
            }
        };
        let max_bytes = input
            .get("maxBytes")
            .and_then(Value::as_u64)
            .unwrap_or(DEFAULT_MAX_BYTES as u64) as usize;
        let file = match fs::File::open(&resolved).await {
            Ok(file) => file,
            Err(error) => {
                return ToolResult::failure(
                    "FILE_READ_ERROR",
                    error.to_string(),
                    context.started_at,
                    json!({"path":path_text}),
                );
            }
        };
        let mut bytes = Vec::with_capacity(max_bytes.min(metadata.len() as usize));
        if let Err(error) = file.take(max_bytes as u64).read_to_end(&mut bytes).await {
            return ToolResult::failure(
                "FILE_READ_ERROR",
                error.to_string(),
                context.started_at,
                json!({"path":path_text}),
            );
        }
        let hash = hex::encode(Sha256::digest(&bytes));
        let content = String::from_utf8_lossy(&bytes);
        let lines = content.split('\n').collect::<Vec<_>>();
        let start = input.get("startLine").and_then(Value::as_u64).unwrap_or(1) as usize;
        let requested_end = input
            .get("endLine")
            .and_then(Value::as_u64)
            .unwrap_or(lines.len() as u64) as usize;
        let end = requested_end.min(lines.len());
        let selected = if start > lines.len() {
            String::new()
        } else {
            lines[start - 1..end].join("\n")
        };
        ToolResult::success(
            json!({
                "path": path_text,
                "content": selected,
                "startLine": start,
                "endLine": end,
                "totalLines": lines.len(),
                "truncated": metadata.len() > max_bytes as u64,
                "sha256": hash,
            }),
            context.started_at,
            json!({"fileSize":metadata.len(),"resolvedPath":resolved}),
        )
    }
}

trait OptionValueExt {
    fn transpose_option(self, name: &str, issues: &mut Vec<String>) -> Option<Option<u64>>;
}

impl OptionValueExt for Option<Option<u64>> {
    fn transpose_option(self, name: &str, issues: &mut Vec<String>) -> Option<Option<u64>> {
        if matches!(self, Some(None)) {
            issues.push(format!("{name} 必须是非负整数。"));
        }
        self
    }
}
